File: DD/mediareader/src/main/java/com/andruid/magic/mediareader/viewmodel/TrackViewModel.java
package com.andruid.magic.mediareader.viewmodel;

import android.app.Application;
import android.os.Bundle;
import android.support.v4.media.MediaBrowserCompat;

import com.andruid.magic.mediareader.data.Constants;
import com.andruid.magic.mediareader.datasourcefactory.TrackDataSourceFactory;
import com.andruid.magic.mediareader.model.Track;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

public class TrackViewModel extends AndroidViewModel {
    private Application application;

    public TrackViewModel(@NonNull Application application) {
        super(application);
        this.application = application;
    }

    public LiveData<PagedList<Track>> getTracks(MediaBrowserCompat mediaBrowserCompat, Bundle options){
        TrackDataSourceFactory dataSourceFactory = new TrackDataSourceFactory(application.getApplicationContext(),mediaBrowserCompat, options);
        PagedList.Config config = new PagedList.Config.Builder()
                .setEnablePlaceholders(false)
                .setPageSize(Constants.PAGE_SIZE)
                .build();
        return new LivePagedListBuilder<>(dataSourceFactory,config).build();
    }
}

File: DiscoDruid/app/src/main/java/com/andruid/magic/discodruid/ui/viewmodel/TrackViewModel.kt
package com.andruid.magic.discodruid.ui.viewmodel

import android.os.Bundle
import android.support.v4.media.MediaBrowserCompat
import androidx.lifecycle.ViewModel
import androidx.lifecycle.asLiveData
import androidx.lifecycle.liveData
import androidx.lifecycle.viewModelScope
import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.cachedIn
import com.andruid.magic.discodruid.data.PAGE_SIZE
import com.andruid.magic.discodruid.paging.TracksPagingSource
import kotlinx.coroutines.cancel

class TrackViewModel(
    private val mediaBrowserCompat: MediaBrowserCompat,
    private val options: Bundle? = null
) : ViewModel() {
    val tracksLiveData = liveData {
        val config = PagingConfig(pageSize = PAGE_SIZE, enablePlaceholders = false)
        val pager = Pager(config) {
            TracksPagingSource(mediaBrowserCompat, options)
        }

        emitSource(
            pager.flow.cachedIn(viewModelScope)
                .asLiveData()
        )
    }

    override fun onCleared() {
        super.onCleared()
        viewModelScope.cancel()
    }
}